atlas.tf

```diff
@@ -0,0 +1,51 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+provider "mongodbatlas" {
+  public_key  = var.atlas_pub_key
+  private_key = var.atlas_priv_key
+}
+
+resource "mongodbatlas_project" "demo" {
+  name   = local.project_id
+  org_id = var.atlas_org_id
+}
+
+resource "mongodbatlas_project_ip_access_list" "acl" {
+  project_id = mongodbatlas_project.demo.id
+  cidr_block = "0.0.0.0/0"
+}
+
+resource "mongodbatlas_cluster" "cluster" {
+  project_id = mongodbatlas_project.demo.id
+  name       = local.project_id
+
+  provider_name               = "TENANT"
+  backing_provider_name       = "GCP"
+  provider_region_name        = "CENTRAL_US"
```

Comment:
This should be parameterized

Reply:
Definitely. I chose to hardcode it now because it was expedient but it definitely needs to be parameterized before we ship.

One challenge is that Atlas uses their own names to refer to our regions, so users will either have to know that and make sure they're putting in the same region for Atlas and Cloud Run, or we could try to manage that for them with a lookup table or some other means of translation. Didn't seem worth it for right now but if you have ideas about how to manage this I'd love to hear them.

Reply:
Great point. I would normally say that it's up to the user to ensure they read the Terraform Registry documentation and ensure they are using a valid region name, but the part that gave me pause was that the Cloud Run region needs to be the same as the Atlas region - I didn't realize that was a requirement (but it makes sense in retrospect). Maybe writing a little helper expression to do some error checking would be wise (or at least adding some comments). Terraform is limited in what we can do within HCL - if we want a full lookup table (kept up to date with an API call), error checking, etc we'll probably have to shell out to go/python/bash with the `local-exec` provisioner. I'd have to think more about how we would do this because even if we went with a statically-linked Golang binary, we'd need a way to use different binaries for different OS/CPU architectures. Maybe just add a comment for now and we can come back to this if we have time.

Reply:
I suppose that technically it doesn't _neeeeeeeeeeed_ to be the same region, but my general position is that most people will want it to be and we should design/optimize for that case while giving folks with more specialized requirements a way to meet their needs.

Having slept on this, I think that any amount of cleverness is going to be too costly and lead to pain down the road, and so I think the best solution (for now, and possibly forever) is strong documentation. What do you think?

Reply:
We have a task (#6) that we were thinking we wouldn't need, but perhaps this is one good use case for a pre-configuration script that lets the user choose from a list of regions that work with Atlas. I say we don't expend much energy on making the MVP too auto-magical for now (just like we assume the images are already pushed to the project). We can prioritize and iterate on enhancements later.

Reply:
@bleything I realize this is a WIP, but ping as soon as you feel you have a working baseline. Given that list of things you enumerated in the PR description, I'd much prefer merging incremental PRs instead of one big monolithic one. Up to you, though.

Reply:
@villasenor @bleything My question might be a bit naive since I'm not familiar with Terraform, so I don't have context on complexity. I also thought about shelling out to the script that I had started, but worth creating a provider (or leveraging a simple [REST provider](https://github.com/Mastercard/terraform-provider-restapi)) to call out to this [Atlas API](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#operation/returnAllCloudProviderRegions) for a list? Probably out of scope for the MVP, but what are your thoughts?

Reply:
@subfuzion ideally there would be a [data source](https://www.terraform.io/language/data-sources) in the atlas provider which allowed us to look up cloud provider regions based on atlas regions (or vice versa). I agree it's out of scope for right now, but if we have someone comfortable with go it might be a nice opportunity to contribute a PR to the atlas provider.

Let's punt on this for now. I'll open an issue for a possible future enhancement.

Reply:
enhancement issue: #8

Reply:
Lol oops. Yes, #9